### PR TITLE
Feature/polygon reducer

### PR DESCRIPTION
**Pull request by Jacks0nJ**
# Polygon extractor and reducers added
This feature branch adds an extractor and reducers for polygon/freehand data.

Relevant documentation and unit tests (100% coverage) have been added.

### Extractor
An extractor for both polygon and freehand data has been added. The output format is lists of lists for both the `x` and `y` data of the polygons drawn.  The `gold_standard`  boolean of the data classification is also extracted.

### Reducers

Two reducers have been added. Both use the [IoU metirc](https://en.wikipedia.org/wiki/Jaccard_index) to define the distance between polygons for clustering with [DBSCAN](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.DBSCAN.html). The data processing is done by converting the x and y data into [shapely polygons](https://shapely.readthedocs.io/en/2.0.6/reference/shapely.Polygon.html).

The main reducer aggregates the polygons of a cluster into a single reduction. As polygons of a cluster can have a different number of sides, standard methods for average (e.g. Euclidean)  cannot be used. Instead, a number of options for averaging the cluster, given by the kwarg `average_type`, have been added.

The "consensus" is found from the average IoU distance between polygons of the same cluster. It is not normalised by the number of volunteers contributing to the cluster.

An additional reducer which finds the contours of intersection/overlap has also been added. E.g. for a cluster, a series of polygons are found, where each polygon is the area of agreement between n volunteers and the next polygon is the area of agreement between n+1 volunteers. This reducer returns one cluster per row, rather than the standard one subject per row. This is to avoid the data structure being too hierarchical. Note it currently takes approximately 100x longer than the standard polygon reducer.

### User guide
A fully functioning jupyter notebook, detailing how to use these new tools for a realistic test project and how to plot the resulting data, has been added to the docs.

### Hide progress bar
The progress bar for the batch extractions and reductions can now be hidden using the option `-hd` on the command line.

### List reductions
Reducers which return lists of dictionaries are now partly supported. This change mostly involved updating the test factory in `base_test_class.py`.

### Time of creation available for reducers
The time when an extraction was created, denoted as `created_at`,  is now provided in the kwargs given to the reducer.

Closes: #787 #792

### CODE REVIEW

File: panoptes_aggregation/reducers/polygon_reducer.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+from .text_utils import tokenize
+import warnings
+import shapely
+
+with warnings.catch_warnings():
+    # collatex is a bit old, we can safely ignore this message as the display
+    # functions are optional and never used in this code
+    warnings.filterwarnings('ignore', category=DeprecationWarning, message='Importing display')
+    import collatex as col
+
```
Comment (CKrawczyk) [inline panoptes_aggregation/reducers/polygon_reducer.py:25]: This is only needed in the text transcription file, I guess it is a holder over from a copy-paste job.

File: panoptes_aggregation/reducers/polygon_reducer.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+DEFAULTS = {
+    'min_samples': {'default': 2, 'type': int},
+    'eps': {'default': 0.5, 'type': float},
+    'average_type': {'default': 'median', 'type': str}
+}
+
+# override the built-in tokenize
+col.core_classes.WordPunctuationTokenizer.tokenize = tokenize
```
Comment (CKrawczyk) [inline panoptes_aggregation/reducers/polygon_reducer.py:33]: Again, not needed in this file.

File: panoptes_aggregation/reducers/polygon_reducer_utils.py [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+        if safety > 10:
+            all_unified = True
+            break
+    areas = [polygons[i].area for i in range(len(polygons))]
+    largest_polygon_index = np.argmax(areas)
+    return polygons[largest_polygon_index]
+
+
+# This needs a list of shapely objects to work!
+def IoU_metric_polygon(a, b, data_in=[]):
```
Comment (CKrawczyk) [inline panoptes_aggregation/reducers/polygon_reducer_utils.py:71]: Can this use the IoU metric function already defined in the regular shape reducer utils?

File: panoptes_aggregation/reducers/polygon_reducer_utils.py [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+
+    Returns
+    -------
+    distances_mean : float
+        The mean of the `IoU_metric_polygon` defined distance between the
+        polygons of the cluster.
+    '''
+    # first check if it is symmetric
+    if not issymmetric(distances_matrix):
+        raise Exception('`distances_matrix` must be a symmetric-square array')
```
Comment (CKrawczyk) [inline panoptes_aggregation/reducers/polygon_reducer_utils.py:185]: Do we really need to check this?  We construct that distance matrix above and know it is symmetric, or am I missing something?

File: panoptes_aggregation/reducers/polygon_reducer_utils.py [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+    # Check the diagonals are all 0, as the distance between an object and its self is always zero
+    if np.sum(np.diagonal(distances_matrix)) != 0:
+        raise ValueError('`distances_matrix` must have zero diagonal elements, as distance between the object and itself is zero')
```
Comment (CKrawczyk) [inline panoptes_aggregation/reducers/polygon_reducer_utils.py:188]: Again, is this needed here?

File: panoptes_aggregation/reducers/polygon_reducer_utils.py [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+            created_at_list.append(datetime.datetime.strptime(time, "%Y-%m-%d %H:%M:%S UTC"))
+    elif isinstance(created_at[0], pdtimestamp):
+        for time in created_at:
+            created_at_list.append(time.to_pydatetime())
+    elif isinstance(created_at[0], datetime.datetime):  # If already correct
+        created_at_list = created_at
+    else:
+        raise Exception('`created_at` needs to contain either UTC strings, pandas timestamps or datetime objects')
+    # sort in time order
+    order_logic = np.argsort(created_at_list)
```
Comment (CKrawczyk) [inline panoptes_aggregation/reducers/polygon_reducer_utils.py:237]: Can you use `argmax` instead?

**Review by CKrawczyk — CHANGES_REQUESTED** [review]
I have tested this locally and was able to run through the new notebook tutorial without issue.  Only a couple of small changes and questions left as inline comments.

File: panoptes_aggregation/reducers/polygon_reducer_utils.py (hunk 5 above)
Comment (CKrawczyk) [reply panoptes_aggregation/reducers/polygon_reducer_utils.py:188]: Not a huge deal, more of a question rather than asking for a change.

File: panoptes_aggregation/reducers/polygon_reducer_utils.py (hunk 3 above)
Comment (Jacks0nJ) [reply panoptes_aggregation/reducers/polygon_reducer_utils.py:71]: The custom metric needs to have this format to get around the limitations scikit-learn puts on the argument to fit with DBSCAN. So we fit for `X` but the actual polygon shapely objects to cluster are stored in `data` and provided in the "metric_params" dictionary to DBSCAN. 

 [`shape_metric_IoU.pyIoU_metric`](https://github.com/zooniverse/aggregation-for-caesar/blob/feature/polygon_reducer/panoptes_aggregation/reducers/shape_metric_IoU.py) would need to be heavily modified to work meet this requirement.

Comment (CKrawczyk) [reply panoptes_aggregation/reducers/polygon_reducer_utils.py:71]: Good point.

**Review by CKrawczyk — APPROVED** [review]
This all looks good to me, happy for you to merge it when you are ready @Jacks0nJ
